### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by nicky-johnstone**
Creating this pull request before it becomes excessively large and will work on another branch for other documentation.

Documentation added/updated:
* Registration
* Filters
  * filter_authenticate
  * filter_validate
* Data Types: is, cast, castable
  * qc::is API
  * qc::cast API
  * qc::castable API
* Added missing proc docs
* Updated index

### CODE REVIEW

**Comment by daniel-qcode** [comment]
@bvw 

Looks good to me

File: doc/filters.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+
+filter_authenticate
+-------------------
+
+The task of this filter is to check if the session and authentication tokens are present and valid and act accordingly.
+
+When this filter is called it will check if the request has been [registered for authentication]. If so then a check is done to determine if the session is valid.
+
+If the session is valid then the request method is checked. If it's not a GET request the authenticity token must be checked for validity. If the token is invalid then an error is returned to the client. Otherwise  the server continues normal execution.
```
Comment (daniel-qcode) [inline doc/filters.md]: ... not a GET or HEAD request the ...

File: doc/filters.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+When a request comes in and this filter is called it will check if the request has been [registered for validation] and if a [request handler] exists to handle the request.
+
+If both of these conditions are met then the relevant data for the request is validated against the data model. If a [custom validation handler] exists for the request then that is called.
+
+If anything failed validation then the [JSON response] is returned to the client otherwise the server continues normal execution.
+
+### Usage
+
+This filter is intended to be used prior to the request being handled so that invalid data isn't passed through to the request handler. This means that `filter_validate` should be registered with Naviserver during `preauth` or `postauth` but never for `trace`.
+
```
Comment (bvw) [inline doc/filters.md]: I'd like to discuss if we should just suggest postauth usage.

File: doc/filters.md [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+part of [Qcode Documentation](index.md)
+
+* * *
+
+Filters, registered on Naviserver with `ns_register_filter`, allow for certain tasks to be carried out at certain stages of Naviserver authorization. See the [Naviserver register docs] for more information.
+
+The qcode-tcl library provides a few useful filters.
+
+
+filter_validate
```
Comment (bvw) [inline doc/filters.md]: qc::filter_validate ?

File: doc/registration.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+
+`validate` is used to register a custom validation handler.
+
+These handlers are used to set up custom validation for a request. They come in useful when validation against the data model perhaps isn't possible or when a developer might want to validate some input differently.
+
+Validation handlers are not set up to authenticate and will only be called if a request handler matching this request is also registered.
+
+### Record Modification
+
+When setting up a custom validation handler it is important to modify the record portion of the [JSON response] with the results of custom validation. If the record is not used appropriately then the client will not get the correct feedback. See the [JSON response API] for help.
```
Comment (bvw) [inline doc/registration.md]: I'd like some more explanation of what happens when validation passes or fails.

**Pull request merged**